### PR TITLE
feat: actions rework

### PR DESCRIPTION
**Pull request by timc1**
Base architecture for actions which enables better testability, and implicit parent/children relationships.

### Background/Problem

Currently, users pass a list of actions to register in kbar:

```
const actions = [
  {
    id: "action1",
    name: "Action 1",
  },
  {
    id: "action2",
    name: "Action 2",
    parent: "action1",
  },
  {
    id: "action3",
    name: "Action 3",
    parent: "action2",
    perform: () => console.log("hello"),
  }
]
```

When they want to register more actions later on, they'll pass another list of actions to `useRegisterActions`:

```
const actions = [
  {
    id: "action4",
    name: "Action 4",
    parent: "action2",'
    perform: () => console.log("yo")
  },
]
```

What we are doing right now under the hood is [implicitly adding a `children` property to each action](https://github.com/timc1/kbar/blob/main/src/useStore.tsx#L52-L82). When we want to remove these actions, we call the closured [`unregister`](https://github.com/timc1/kbar/blob/main/src/useStore.tsx#L84-L106).

While all this is fine, the logic here is difficult to test, and extending behavior; e.g. adding a reference of all ancestors in addition to children, will start making this setState complicated.

Additionally, we should not expose the `children` property of `Action` to the end user, as they only need to ever define a `parent` to represent a nested action.

### This solution

We introduce two objects – `ActionInterface` and `ActionImpl` – which will manage the implicit parent/child relationships between all actions, in addition any other relationships we want; e.g. tracking ancestors.

#### `ActionImpl`

An implementation of the `Action` type. Users pass a list of actions in the example above, and we transform each action into an `ActionImpl` which will keep a reference of its `parent`, `children`, and `ancestors`. As actions are added and removed, this class will handle updating its references.

#### `ActionInterface`

A utility which help us interface with the underlying `ActionImpl` class. This holds the internal `actions` state which our `useStore` will use.

### CODE REVIEW

File: src/action/ActionImpl.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import { ReactElement, JSXElementConstructor, ReactNode } from "react";
+import type { BaseAction, Action } from "../types";
+
+interface ActionImplOptions {
+  parent?: ActionImpl;
+}
+
+export class ActionImpl implements Action {
```
Comment (timc1) [inline src/action/ActionImpl.ts:8]: Users define actions by passing in an object like so:

```
{
  id: 123,
  name: "Post 1",
  parent: "parentId"
}
```

and we transform that into our internal `ActionImpl` object, which will maintain `parent`, `children`, and `ancestor` relationships outside of React land.

File: src/action/ActionInterface.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import type { BaseAction, ActionId } from "../types";
+import { ActionImpl } from "./ActionImpl";
+
+export class ActionInterface {
+  actions: Record<ActionId, ActionImpl> = {};
```
Comment (timc1) [inline src/action/ActionInterface.ts:5]: This interface improves the previous naive logic we had for adding/removing actions. This way is more testable and extracts logic out of our React component.

**Comment by timc1** [comment]
@tommoor been thinking about how we can make our actions more robust and testable. lmk what you think 🙏

File: src/useStore.tsx [hunk 3]
```diff
@@ -49,58 +48,19 @@ export default function useStore(props: useStoreProps) {
...
-            parent.children = parent.children.filter(
-              (child) => child !== actionId
-            );
-          }
-          delete allActions[actionId];
-        });
         return {
           ...state,
-          actions: allActions,
+          actions: actionsInterface.current.remove(actions),
```
Comment (tommoor) [inline src/useStore.tsx:63]: Certainly way cleaner to reason about here

File: src/InternalEvents.tsx [hunk 4]
```diff
@@ -177,7 +177,7 @@ function useShortcuts() {
         }
         if (action.shortcut.join("") === bufferString) {
           event.preventDefault();
-          if (action.children) {
+          if (action.children?.length) {
```
Comment (tommoor) [inline src/InternalEvents.tsx:180]: 👍  I worked around this

**Review by tommoor — APPROVED** [review]
love it

**Pull request merged**